src/snowflake/snowpark/_internal/ast_utils.py

```diff
@@ -346,6 +346,28 @@ def build_expr_from_snowpark_column_or_python_val(
         pass
     else:
         raise TypeError(f"{type(value)} is not a valid type for Column or literal AST.")
+    
+
+def build_expr_from_snowpark_column_or_table_fn(
```

Comment:
why not directly use if and the ` build_expr_from_snowpark_column(expr_builder, value)` helper? Why do we need to introduce a new helper function here?

Reply:
This is to accommodate arguments to `with_column(s)` which only allow Snowpark `Column` instances or `TableFunctionCall` instances in the `values` parameter.

I created this helper in anticipation of the fact that the `TableFunctionCall` ASTs will be generated in the same pattern as `Column` ASTs and stored in an internal `_ast` field. For now we don't support that so the `TableFunctionCall` case will raise a `NotImplementedError`, but later it'll be simple call to `CopyFrom`.